src/sinks/aws_s3/sink.rs

```diff
@@ -0,0 +1,521 @@
+                            .map(|dt| dt.with_timezone(&Utc)),
+                        _ => None,
+                    }),
+                // TODO: We don't ship metrics to the S3, but if we did, would this be right? or is
+                // there an actual field we should be checking similar to above?
+                Event::Metric(_) => Some(Utc::now()),
+            })
+            .unwrap_or_else(|| Utc::now());
+        let formatted_ts = last_event_ts.format(&time_format);
+        */
+        let formatted_ts = Utc::now().format(options.filename_time_format.as_str());
+
+        if options.filename_append_uuid {
+            let uuid = Uuid::new_v4();
+            format!("{}-{}", formatted_ts, uuid.to_hyphenated())
+        } else {
+            formatted_ts.to_string()
+        }
+    };
+
+    let extension = options
+        .filename_extension
+        .as_ref()
+        .cloned()
+        .unwrap_or_else(|| options.compression.extension().into());
+    let key = format!("{}/{}.{}", key, filename, extension);
+
+    // Process our events. This does all of the necessary encoding rule
+    // application, as well as encoding and compressing the events.  We're
+    // handed back a tidy `Bytes` instance we can send directly to S3.
+    //
+    // TODO: we need to do something with these
+    let batch_size = batch.len();
+    let (body, finalizers) = process_event_batch(batch, &options.encoding, options.compression);
+
+    debug!(
+        message = "Sending events.",
+        bytes = ?body.len(),
+        bucket = ?options.bucket,
+        key = ?key
+    );
+
+    S3Request {
+        body,
+        bucket: options.bucket.clone(),
+        key,
+        content_encoding: options.compression.content_encoding(),
+        options: options.api_options.clone(),
+        batch_size,
+        finalizers,
+    }
+}
+
+pub fn process_event_batch(
+    batch: Vec<Event>,
+    encoding: &EncodingConfig<Encoding>,
+    compression: Compression,
+) -> (Bytes, EventFinalizers) {
+    enum Writer {
+        Plain(Vec<u8>),
+        GzipCompressed(GzEncoder<Vec<u8>>),
+    }
+
+    impl Write for Writer {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            match self {
+                Writer::Plain(inner_buf) => inner_buf.write(buf),
+                Writer::GzipCompressed(writer) => writer.write(buf),
+            }
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            match self {
+                Writer::Plain(_) => Ok(()),
+                Writer::GzipCompressed(writer) => writer.flush(),
+            }
+        }
+    }
+
+    // Build our compressor first, so that we can encode directly into it.
+    let mut writer = {
+        // This is a best guess, because encoding could add a good chunk of
+        // overhead to the raw, in-memory representation of an event, but if
+        // we're compressing, then we should end up net below the capacity.
+        let buffer = Vec::new();
```

Comment:
The comment makes it sound like we should be using `Vec::with_capacity` here.

Reply:
Indeed. Right now it's a best-guess of 0.

Reply:
Bumped in 69a60138f

Reply:
Agreed.  The code was `with_capacity` when I wrote it, because I had access to the raw size, in bytes, of the events in the batch.  Our more generic code for `Batcher` doesn't currently expose that, so I had to change to the line as it stands, but left the comments with the plan of going back and augmenting the batch type to expose the total bytes.

Reply:
That's an excellent idea. Do you have an issue open for it already?